### ISSUE TITLE
Debugging hangs when Stack window includes long data elements

### ISSUE THREAD

**Issue by slanska** [pkulchenko/ZeroBraneStudio#863]
After upgrading from 1.60 to 1.70, I started experiencing ZeroBrane freezing during debugging session. Symptoms are similar to when I am using EmmyLua plugin for IntelliJ IDEA family (I am using CLion, to be exact). I am running MacBook Pro 2015. Since it is barely usable in my case, I would like to revert to 1.60 until the issue is resolved. Where can I find earlier builds of ZB?

**Comment by slanska** [comment]
[error_report.txt](https://github.com/pkulchenko/ZeroBraneStudio/files/1662547/error_report.txt)

**Comment by slanska** [comment]
UPDATE: After little probing, I found out that hanging in ZeroBrane (and also sluggishness in EmmaLua plugin) is somehow related to the data size I am trying to process. Essentially, my program loads external JSON content (~1.6 MB), parses it to a Lua table using cjson decode, and then starts processing. Once I switched to smaller data file (few kilobytes), debugging again became fast and smooth. Based on post mortem dump from frozen ZeroBrane (attached in the comment above), the issue might be related to LuaSocket, which is used by mobdebug.

**Comment by pkulchenko** [comment]
@slanska, to answer your questions: all the previous versions are available in the same folder as the current version; just replace the version number in the download link (from 1.70 to 1.60 or 1.50 and so on).

In terms of the debugging performance, there hasn't been any significant changes to the IDE or the debugger that may be contributing to the slowdown. Is it possible that you switched to using larger data structures about the same time you switched to 1.70?

> Based on post mortem dump from frozen ZeroBrane (attached in the comment above), the issue might be related to LuaSocket, which is used by mobdebug.

There hasn't been any changes to luasocket in the last several releases, so it's probably not the reason.

Do you have the Stack window open during debugging by any chance? The size of the data structures shouldn't have any impact on the IDE itself, unless you have a stack window open, which will send the content of the local value and upvalues to the IDE after each step. I don't think there are any other scenarios that would trigger encoding of those JSON structures during debugging.

If it's indeed caused by the content being sent back to the IDE, you may want to re-configure `debugger.maxdatanum` and `debugger.maxdatalevel` to smaller values (as described in the [debugger section](https://studio.zerobrane.com/doc-general-preferences#debugger)).

If you do *not* have the Stack window open and there is indeed a noticeable difference between the performance you see in 1.70 and 1.60 versions, I'd very much like to see if you can bisect the changes between 1.60 and 1.70 to identify the update that might have caused this. This will only require 7 steps and I can help you with the process; if you can quickly identify if the selected commit is "good" or "bad", it should take less than 15m.

**Comment by slanska** [comment]
After playing with 1.60 and 1.70 I cannot see any difference during debugging - both versions hang on attempt to debug certain portion of code. The issue is definitely related to amount of data to process - since issue disappeared once I switched to smaller payload. I did have stack window opened, though I dont’t think I had any variables there. I can try to confirm that. I will play with debug settings you recommended to see how it would impact debugging. And thanks for the prompt reply.

**Comment by pkulchenko** [comment]
If you think you can reduce it to a small example that works with that json data, I can try to reproduce locally to see if I can troubleshoot it.

**Comment by slanska** [comment]
I will try to prepare a small project later today.

**Comment by slanska** [comment]
Attached please find a small demo which freezes ZeroBrane and makes EmmyLua (plugin for IntelliJ) very sluggish. 
[slow_debug_demo.zip](https://github.com/pkulchenko/ZeroBraneStudio/files/1666358/slow_debug_demo.zip)

**Comment by pkulchenko** [comment]
@slanska, thank you for the sample! I think it's related to the photo elements more than anything else, as the stack window will try to fix UTF8 encoding before showing the content and given that the photos are long enough (50k+), it takes a long time to scan and fix them. I added trimming to maxlen already specified for other fields, which should address this issue.

Can you try the following patch?

```diff
diff --git a/src/editor/debugger.lua b/src/editor/debugger.lua
index f7ba7dc..97e075a 100644
--- a/src/editor/debugger.lua
+++ b/src/editor/debugger.lua
@@ -144,6 +144,7 @@ function debugger:updateStackSync()
 
     local forceexpand = ide.config.debugger.maxdatalevel == 1
     local params = debugger:GetDataOptions({maxlevel=false})
+    local maxlen = tonumber(ide.config.debugger.maxdatalength)
 
     local root = stackCtrl:AddRoot("Stack")
     callData = {} -- reset call cache
@@ -181,6 +182,7 @@ function debugger:updateStackSync()
         -- format the variable name, value as a single line and,
         -- if not a simple type, the string value.
         local value = val[1]
+        if type(value) == "string" and maxlen and #value > maxlen then value = value:sub(1,maxlen) end
         local text = ("%s = %s"):format(name, fixUTF8(serialize(value, params)))
         local item = stackCtrl:AppendItem(callitem, text, image.LOCAL)
         stackCtrl:SetItemValueIfExpandable(item, value, forceexpand)
@@ -190,6 +192,7 @@ function debugger:updateStackSync()
       -- add the upvalues for this call stack level to the tree item
       for name,val in pairs(type(frame[3]) == "table" and frame[3] or {}) do
         local value = val[1]
+        if type(value) == "string" and maxlen and #value > maxlen then value = value:sub(1,maxlen) end
         local text = ("%s = %s"):format(name, fixUTF8(serialize(value, params)))
         local item = stackCtrl:AppendItem(callitem, text, image.UPVALUE)
         stackCtrl:SetItemValueIfExpandable(item, value, forceexpand)
```

**Comment by slanska** [comment]
Your fix helped! Debugging is working ok now. Спасибо, Павел.

**Comment by pkulchenko** [comment]
ok; great! I'll get it in then. Thank for checking.

**Comment by vanpersie32** [comment]
@pkulchenko core dump when start zerobrane studio from script zbstudio.sh

**Comment by pkulchenko** [comment]
@vanpersie32, you mean on the current master? Do you think it's related to the changes for this ticket?

**Comment by pkulchenko** [comment]
If yes, can you try on an earlier version or bisect the changes that cause the crash?

Also, can you try with an empty setup by removing the [session configuration file](https://studio.zerobrane.com/doc-general-preferences#session-configuration) (you may save a copy in case you need to restore your setup).

**Comment by vanpersie32** [comment]
@pkulchenko thank you, the later approach  works for me. I really appreciate your reply.

**Comment by pkulchenko** [comment]
@vanpersie32, you mean it's back to normal after you restored the session config file? Did you save the one that was causing the problem by any chance?

**Comment by vanpersie32** [comment]
@pkulchenko  I remove the .ZeroBraneStudio in the Home directory. Then I start Zerobrane studio by sh, It is ok now

**Comment by pkulchenko** [comment]
If the crash happens again, please save the session config file before removing and send it to me (the email is in my github profile).

**Comment by slanska** [comment]
@pkulchenko UPDATE: After I applied your fix , I tested only sample application (slow_debug_demo.lua), with single Lua script. It works OK. However, when I run my main script, on the same JSON data file, it still hangs. What any additional info I could provide you with for help? Thanks.

**Comment by pkulchenko** [comment]
Is the JSON data the same in both cases?

Can you try adding the following snippet to the config file? This will strip all Photo elements from the serialized output sent by the debugger. At least we will know if the slowdown is related to the Photo elements or to something else.

```lua
debugger.init = [[
            local mdb = require('mobdebug')
            local dump = mdb.dump
            mdb.dump = function(v, opts)
                opts = opts or {}
                opts.keyignore = {Photo = true}
                return dump(v, opts)
            end
            local line = mdb.line
            mdb.line = function(v, opts)
                opts = opts or {}
                opts.keyignore = {Photo = true}
                return line(v, opts)
            end
    ]]
```

**Comment by slanska** [comment]
@pkulchenko yes, it is the same file. Sure, I will try your snippet. Thanks!

**Comment by slanska** [comment]
@pkulchenko my apologies for delayed reply - was busy with other errands for the last few days.
I am not sure where apply your snipper - I tried in a) user settings (/.zbstudio/user.lua) and b) config.lua (zbstudio/config.lua) in ZB package directly. Neither helped with the issue. Am I applying your snippet it correctly? Thanks.

**Comment by pkulchenko** [comment]
@slanska, yes, it should be in user.lua; you can open it with `Edit | Preferences | Settings: User`.

> However, when I run my main script, on the same JSON data file, it still hangs. What any additional info I could provide you with for help?

Can you provide the script I can test it on? Or at least update the script you provided earlier to make the issue reproducible on it?

**Comment by slanska** [comment]
@pkulchenko I tried to debug my app again and now it seems to be OK. Not sure what has changed since I tried it last time. Anyway, if you'd like to check peculiarities of debugging on specific code, here is my project - https://github.com/slanska/flexilite. Run src_lua/test_flexi.lua, breakpoint can be placed in, for instance, src_lua/flexi_DataUpdate.lua:179. Before this breakpoint would cause debugging to freeze 100%. Thank you for your help!
